src/Features/Core/Portable/ConvertForToForEach/AbstractConvertForToForEachCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,468 @@
+            if (iterationSymbol != null)
+            {
+                if (iterationSymbol.Locations.Length != 1 ||
+                    !iterationSymbol.Locations[0].IsInSource ||
+                    iterationVariable != iterationSymbol.Locations[0].FindToken(cancellationToken))
+                {
+                    // was a reference to some other variable.
+                    return;
+                }
+            }
+
+            // Make sure we're starting at 0.
+            var initializerValue = semanticModel.GetConstantValue(initializer, cancellationToken);
+            if (!(initializerValue.HasValue && initializerValue.Value is 0))
+            {
+                return;
+            }
+
+            // Make sure we're incrementing by 1.
+            if (stepValueExpressionOpt != null)
+            {
+                var stepValue = semanticModel.GetConstantValue(stepValueExpressionOpt);
+                if (!(stepValue.HasValue && stepValue.Value is 1))
+                {
+                    return;
+                }
+            }
+
+            var collectionType = semanticModel.GetTypeInfo(collectionExpression, cancellationToken);
+            if (collectionType.Type == null && collectionType.Type.TypeKind == TypeKind.Error)
+            {
+                return;
+            }
+
+            var containingType = semanticModel.GetEnclosingNamedType(context.Span.Start, cancellationToken);
+            if (containingType == null)
+            {
+                return;
+            }
+
+            var ienumerableType = semanticModel.Compilation.GetSpecialType(SpecialType.System_Collections_Generic_IEnumerable_T);
+            var ienumeratorType = semanticModel.Compilation.GetSpecialType(SpecialType.System_Collections_Generic_IEnumerator_T);
+
+            // make sure the collection can be iterated.
+            if (!TryGetIterationElementType(
+                    containingType, collectionType.Type,
+                    ienumerableType, ienumeratorType,
+                    out var iterationType))
+            {
+                return;
+            }
+
+            // If the user uses the iteration variable for any other reason, we can't convert this.
+            var bodyStatements = GetBodyStatements(forStatement);
+            foreach (var statement in bodyStatements)
+            {
+                if (iterationVariableIsUsedForMoreThanCollectionIndex(statement))
+                {
+                    return;
+                }
+            }
+
+            // Looks good.  We can convert this.
+            context.RegisterRefactoring(new MyCodeAction(GetTitle(),
+                c => ConvertForToForEachAsync(
+                    document, forStatement, iterationVariable, collectionExpression,
+                    containingType, collectionType.Type, iterationType, c)));
+
+            // local functions
+            bool iterationVariableIsUsedForMoreThanCollectionIndex(SyntaxNode current)
```

Comment:
lowercase? is this the style we use?

Reply:
i personally prefer it.  It's just a variable.  If it were a local lambda we'd lowercase it.  So being a local function just keeps that style.  there is no strict rule here.

Reply:
#bydesign

Reply:
is current statement? or multiple kind node?